### DISCUSSION TITLE
Can SSCHA code deal with 2D materials?

### DISCUSSION THREAD

**Discussion by nintendo-64** [SSCHAcode/python-sscha#44] [Q&A] [answered]
Dear developers,

Recently, I have analyzed the anharmonic effect in 2D materials doped graphene. 100 configurations are generated at 100K. During structure-fixed minimization, the error bar is much larger than FC gradient and only runs one step, so I change the meaningful_factor to 0.1 while keeping kong_liu_ratio=0.5 and min_step_dyn=0.05. The final FC gradient decreases to 500 and the error increases to 5000. I want to examine the convergence but do not know how to use the &utils namespace in python script. Then I enlarge the number of configurations to 200 and 300. The final gradient and error decrease to 300 and 3000 for 200 configurations. However, for 300 configurations, the final gradient and error are much larger with about 1000000 and 50000. Only one population are calculated during all the calculation. Then I have few questions:

1. How to examine the convergence of frequencies using python script during minimization?
2. Why the gradient increases a lot when generating more configurations? 
3. Can SSCHA code deal with 2D materials?

It will be greatly appreciated if you could spare time to answer my questions.

Yours
Lu Qing

**Comment by mesonepigreco** [comment, marked as answer]
Hi,
You can decrease meaningful_factor to much lower values (e.g 0.0001 or even lower) to further reduce the gradient.
To print the frequencies with the python script you can do in the following way.
You need to tell the code where to save the frequencies, as:

```
import sscha.Utilities

# Setup the saving frequencies
save_freqs = sscha.Utilities.IOInfo()
save_freqs.SetupSaving("frequencies.dat")
    
# Run the minimization
minim.run(custom_function_post = save_freqs.CFP_SaveFrequencies)
```

A full working example of script is the following


```
import sscha, sscha.Ensemble
import sscha.SchaMinimizer
import sscha.Utilities

dyn = CC.Phonons.Phonons(DYN_NAME, NQIRR)
dyn.Symmetrize()
ensemble = sscha.Ensemble.Ensemble(dyn, Tg, dyn.GetSupercell())
ensemble.load(DATA_DIR, POPULATION, N_RANDOM)

minim = sscha.SchaMinimizer.SSCHA_Minimizer(ensemble)
minim.min_step_dyn = MIN_STEP_DYN
minim.min_step_struc = MIN_STEP_STRUC
minim.kong_liu_ratio = KONG_LIU_RATIO

minim.init()
minim.print_info()

# Here the part to save the frequencies   
save_freqs = sscha.Utilities.IOInfo()
save_freqs.SetupSaving("frequencies.dat")
    
# Run the minimization
minim.run(custom_function_post = save_freqs.CFP_SaveFrequencies)
minim.finalize()
minim.dyn.save_qe("final_dyn")
```

The fact that the gradient increases so much with 300 configurations is indeed very strange. What is the supercell you are using? If you start with a very big super cell (for example 10x10x1)  it is possible that 300 configurations are not enough to converge. In this case I suggest starting with smaller cells (e.g. a 4x4x1) and then you can interpolate the result to a finer mesh  and use that one as a starting point. For graphene, the 4x4x1 supercell should have 32 atoms and be very quick to converge, then try to increase to a 6x6x1, 8x8x1 and check how the frequencies or the phonon dispersion changes.

To aswer to your last question, yes, the code can deal with 2D materials, we also employed it on graphene, (see [this one](https://arxiv.org/abs/2005.12047) , for example)

Bests,
Lorenzo

**Reply by nintendo-64** [reply]
Dear Lorenzo,

My original system has 10 atoms ( 2 atoms are added into 2x2x1 graphene) and nqs for phonon calculation is 4x4x1, which means the supercell has 160 atoms. How many generations should I generate at least? To increase the supercell, does that mean calculating phonon dispersions from scratch? 

By the way, I notice that how to check the convergency using python script has been added in FAQs on website and thank you for your help.

Yours,
Lu Qing

**Reply by ionerrea** [reply]
Dear Lu Qing, 

In order to determine the number of generations and the number of configurations needed to converge you need to do the calculations. The ideal is to start from not many configurations and reach a situation (after several populations) where you get the gradients very small with a very small meaningful_factor and a large Kong-Liu ratio (see FAQ http://sscha.eu/faqs/#How-do-I-check-if-my-calculations-are-well-converged?). This would mean that your auxiliary frequencies and centroid positions don't move anymore in the minimization and your statistical sampling is good.

Then you should increase the number of configurations and make sure your result is converged.

Bests,

Ion

**Reply by nintendo-64** [reply]
Dear Professor Ion,

Thank you for your guidance! 

Yours,
Lu Qing
